test_rmw_implementation/test/test_publisher.cpp

```diff
@@ -0,0 +1,245 @@
+  ASSERT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, topic_name, nullptr, &options);
+  ASSERT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_unknown, &options);
+  ASSERT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, nullptr);
+  ASSERT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  // Creating and destroying a publisher still succeeds.
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_publisher(node, pub);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestPublisher, RMW_IMPLEMENTATION), destroy_with_bad_arguments) {
+  rmw_publisher_options_t options = rmw_get_default_publisher_options();
+  const char * const topic_name = "/test";
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  rmw_publisher_t * pub =
+    rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+
+  // Destroying publisher with invalid arguments fails.
+  rmw_ret_t ret = rmw_destroy_publisher(nullptr, pub);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret);
+  rmw_reset_error();
+
+  ret = rmw_destroy_publisher(node, nullptr);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret);
+  rmw_reset_error();
+
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  ret = rmw_destroy_publisher(node, pub);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, ret);
+  rmw_reset_error();
+
+  // Destroying publisher still succeeds.
+  ret = rmw_destroy_publisher(node, pub);
+  EXPECT_EQ(RMW_RET_OK, ret);
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestPublisher, RMW_IMPLEMENTATION), get_actual_qos_from_system_defaults) {
+  rmw_publisher_options_t options = rmw_get_default_publisher_options();
+  const char * const topic_name = "/test";
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  rmw_publisher_t * pub =
+    rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_system_default, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_unknown;
+  rmw_ret_t ret = rmw_publisher_get_actual_qos(pub, &qos_profile);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  EXPECT_NE(rmw_qos_profile_system_default.history, qos_profile.history);
+  EXPECT_NE(rmw_qos_profile_unknown.history, qos_profile.history);
+  EXPECT_NE(rmw_qos_profile_system_default.reliability, qos_profile.reliability);
```

Comment:
The publisher is created with `rmw_qos_profile_system_default` then why when reading the `qos_profile` do not match to this? It seems odd, mind asking a comment on this?

Reply:
I can't see the difference between this section vs the last test case `get_actual_qos`

Reply:
> The publisher is created with rmw_qos_profile_system_default then why when reading the qos_profile do not match to this? It seems odd, mind asking a comment on this?

`_SYSTEM_DEFAULT` are sentinel values, expect to be replaced by concrete policies. This test verifies that's the case in a generic way (i.e. can't test for a specific policy as it is implementation dependent).

> I can't see the difference between this section vs the last test case get_actual_qos

That other test uses concrete policies and so it expects them to be honored.

Just as an FYI, liveliness, lifespan and deadline QoS policies are less consistent in this regard and thus why they are currently not verified. I plan to come back to them at a later point in time.

Reply:
This test probably deserves a comment with this explanation.